### ISSUE TITLE
Error: The process cannot access the file because it is being used by another process.

### ISSUE THREAD

**Issue by Overhatted** [facebook/buck2#436]
When I make a build on Windows I got this error pretty frequently:

Action failed: root//src/MainLib:MainLib (cxx_compile File.cpp)
Internal error (stage: materialize_inputs_failed): Error cleaning up output path `buck-out/v2/tmp/root/fb50fd37ce946800/src/MainLib/__MainLib__/cxx_compile/File.cpp`: remove_path_recursive(E:\Project\buck-out\v2\tmp\root\fb50fd37ce946800\src\MainLib\__MainLib__\cxx_compile\File.cpp) on directory: remove_dir_all(E:\Project\buck-out\v2\tmp\root\fb50fd37ce946800\src\MainLib\__MainLib__\cxx_compile\File.cpp): The process cannot access the file because it is being used by another process. (os error 32)
stdout:
stderr:

It didn't happen when I ran buck2 with a low "--num-threads" or when I reproduced the commands in "buck2 log what-ran" using a Python script so I thought it was some filesystem race-condition.

I added some logs to here https://github.com/facebook/buck2/blob/4e5e76998a3cb0636e415b047798dfdc1ac56c05/app/buck2_execute_impl/src/executors/local.rs#L418 and here https://github.com/facebook/buck2/blob/4e5e76998a3cb0636e415b047798dfdc1ac56c05/app/buck2_execute_impl/src/executors/local.rs#L444 and could see that the files that were causing the error above, were beginning the execution but not ending.

I then added the below code to here https://github.com/facebook/buck2/blob/4e5e76998a3cb0636e415b047798dfdc1ac56c05/app/buck2_forkserver/src/run/mod.rs#L333 and the error stopped happening.

let _ = Command::new("cmd.exe")
    .arg("/c")
    .arg("TASKKILL /F /T /PID ".to_owned() + &child.id().unwrap_or(0).to_string())
    .output();

Basically buck2 should kill the task's process and all its children rather than the current behavior of only killing the parent process. I wasn't able to code a good implementation of killing all the child processes though.

A related question is: why would Buck2 cancel a task?

**Comment by ndmitchell** [comment]
Are you using remote execution? It often races remote vs local execution (to see which one happens first), and then cancels the slower one. If you aren't then it would be a bit weird.

On Windows I think the standard behaviour to kill all children is create the process with `CREATE_NEW_PROCESS_GROUP` and kill it with `GenerateConsoleCtrlEvent` and `CTRL_BREAK_EVENT` (at least, that's what Haskell does). I suspect Rust has that wrapped up in a library somewhere?

**Comment by ndmitchell** [comment]
Roping in our Windows experts @lmvasquezg @KapJI

**Comment by Overhatted** [comment]
I'm not using remote execution or caching. I guess that's something else to investigate and report on another thread.

**Comment by ndmitchell** [comment]
@Overhatted - are you running into an error? When we see one error, we cancel all inflight jobs.

**Comment by KapJI** [comment]
Thanks for debugging this issue!

Yes, it seems that `std::process::Child::kill` only calls `TerminateProcess` which doesn't kill the whole process tree. And Windows doesn't provide a convenient API for that. So instead we should find all processes to kill ourself e.g. using `CreateToolhelp32Snapshot` call.

I can try to fix it next week but if you can submit a PR, that'd be great.

**Comment by Overhatted** [comment]
@ndmitchell The tasks being canceled happen also when the code is correct and compilable.

**Comment by zjturner** [comment]
If you want to kill the whole process tree you can add them to a Windows Job object, and kill the job.

**Comment by Overhatted** [comment]
I'm working on this at the moment although I'm leaning towards CreateToolhelp32Snapshot since tokio might end up supporting process groups in the future through its API: https://docs.rs/tokio/latest/tokio/process/struct.Command.html#method.process_group
I'm going to call CreateToolhelp32Snapshot in a loop until no more child processes exist so I don't think the solution will be racy.

**Comment by zjturner** [comment]
I don't see any indication from that link that they might eventually support Windows.  Note that `CreateToolhelp32Snapshot` *still* has other problems.  A parent can die before its children or its children can die before its parent, naturally without being force killed.  You could imagine a situation where you do this:

1. Call CreateToolhelpSnapshot32
2. Before you finish iterating and killing the processes, one of the processes in the list spawns a child, A.
3. A spawns another child B and then dies, but B keeps running.

How do you know to kill B the next time you call CreateToolhelp32Snapshot?

The only actually correct and reliable way to kill an entire process tree in Win32 is with a Job Object.

**Comment by zjturner** [comment]
One additional data point.  Ninja uses CREATE_NEW_PROCESS_GROUP.  I wouldn’t call it common, but I’ve seen it quite a few times where after I Ctrl+C a ninja build on windows, there are several processes hanging around in Process Explorer highlighted gray.  Usually it’s python or some subprocess that python spawned, or git or one of its processes.


CreateToolhelp32Snapshot has another issue in that windows recycles pids quickly.  So you can have a process in your snapshot saying its parent has PID X, and X is a currently running process which is not actually the parent.  This isn’t even that uncommon, it will happen in practice a lot.

I can’t recommend strongly enough to please if you’re going to fix this, use a job object.  It’s not difficult and it is guaranteed correct

**Comment by Overhatted** [comment]
@zjturner Thanks for your input. I ended up doing the changes with job objects. It was simple to implement and much more complete.

Though Microsoft says "Child processes created using [Win32_Process.Create](https://learn.microsoft.com/en-us/windows/win32/cimwin32prov/create-method-in-class-win32-process) are not associated with the job." which seems to imply that you can create breakaway processes even without the breakaway flags in the job. A bit strange but for the Buck2 use case it probably doesn't matter.

**Comment by zjturner** [comment]
Where are you seeing that?  I'm seeing that it says:

> Processes created with the Win32_Process.Create method are limited by the job object unless the CREATE_BREAKAWAY_FROM_JOB flag is specified

That reads to me as though it *is* part of the job, and you have to specify `CREATE_BREAKAWAY_FROM_JOB` for it to not be part of the job.  Which I'm not sure why they'd even mention, because you would expect that to be the default since it's the standard behavior of `::CreateProcess`.

**Comment by zjturner** [comment]
Huh, interesting 🤷
